highfidelity/hifi
# Secondary camera and rendering improvements

This PR makes the following changes to spectator-camera:
1. ~Added a `SECONDARY_CAMERA_RENDER_MODE` to the `RenderArgs::RenderMode` enum, which doesn't do anything right now, but may be useful in the future~ Removed
2. Correctly set the projection of the secondary camera
    - ~None of the projection's parameters (vFoV, near clip distance, far clip distance) are available to JS _yet_; they are hardcoded right now.~ These parameters are now available to JS.
3. The secondary camera no longer uses the same culling frustum as the main camera.
    - This fixes most culling issues. I have seen some minor culling issues at the edges of the texture when in HMD mode, and I think it's because of the way we only use one eye's culling frustum? Not sure.
4. Fixed a bug in GLBackendTransform.cpp where the stereo enabled status wasn't checked properly
5. Removed dependencies on the `FramebufferCache` singleton for determining texture dimensions, instead correctly relying on viewport dimensions
6. Improved the scaling of the Spectator Camera Viewfinder Image3DOverlay when window size changes
    - "Improved", but not "perfected" - the new code will stretch the preview as the window aspect ratio changes, which should be fixed. I created a Trello card for this.
7. Removed the `flip()` helper function from `spectatorCamera.js`. Investigated and commented why a negative dimension for the viewfinder overlay is necessary.
    - This warrants more investigation when we have time.
8. Removed dead comments and code

At @samcake's request I will make another PR with changes from (4) and (5) to merge into Master.

Huge props to Sam for helping me out with bugfixes I never would have found on my own :)

**Test Plan:**
HMD + Hand Controllers Required
1. Don your HMD and navigate to dev-welcome. Verify that the "Display->Disable Preview" menu option is CHECKED. Also ensure that the Interface application window is maximized (fullscreen is unnecessary).
2. Using your Tablet, open the "Spectator" app
3. Tap the "Camera On" checkbox
4. Verify that a 3D representation of a video camera appears in front of you at a comfortable distance and height.
5. Verify that there exists a "viewfinder" overlay on top of the camera model that allows you to see things from the camera's point of view.
6. Using a hand controller, grab the camera and move it around. Verify that you are able to move it around.
7. Verify that, as you move the camera around, the viewfinder displays what you would expect the camera to capture i.e. if you grab rotate the camera 45 degrees to the left, the viewfinder shows content to your left.
8. Verify that, as you move the camera around, content does not "pop-in" on the viewfinder.
9. Point the camera's lens towards your avatar. Verify that all parts of your avatar's body and attachments are visible as you would expect.
    - KNOWN ISSUE: Head is cauterized and will not display. See https://trello.com/c/oktxxpnY/185-what-to-do-about-cauterized-head
10. Peek behind your HMD and manually resize the Interface window to be a square. Put your HMD back on. Verify that the camera's viewfinder displays a square image.
    - KNOWN ISSUE: Viewfinder preview will not automatically resize when window is resized. For now, use the tablet app to toggle the camera off and on, then verify the above, with the caveat that...
    - KNOWN ISSUE: Viewfinder preview currently stretches to fit the "glass" piece of the camera's model. For details regarding both this and the above KNOWN ISSUES, see: https://trello.com/c/dqtIcns3/170-respond-to-resizing-screen
11. Open the Spectator tablet app. Verify that the "MONITOR SHOWS" slider is currently set to its default, "HMD Preview".
12. While keeping the Interface window a square, use the "Display" menu to UNCHECK the "Disable Preview" option.
13. As you look around in your HMD, peek behind your HMD to verify that what shows up on your monitor (in the Interface application window) is a preview of what you're looking at.
14. In the Spectator tablet app, tap the "MONITOR SHOWS" slider to move the slider from "HMD Preview" to "Camera View".
15. Peek behind your HMD to verify that what's showing up on your monitor (in the Interface application window) is what the Spectator Camera is "seeing". Ensure the display is not corrupted in any way.
    - KNOWN ISSUE: Lighting problem - you might see "lines" or "tiles" of incorrect lighting on the display. See: https://trello.com/c/WUeqovYH/187-lighting-issue-vertical-line-about-2-3-across-to-the-right
16. Peek behind your HMD and manually resize the Interface window to be maximized (fullscreen is unnecessary). Put your HMD back on. Verify that the camera's viewfinder displays a wide image that fills the "glass" part of the camera model. Also, peek behind your HMD to verify that what displays on your monitor (in the Interface application window) is what the Spectator Camera is "seeing". Ensure the display is not corrupted in any way. 
    - KNOWN ISSUES: same as (10)'s KNOWN ISSUES
    - KNOWN ISSUE: What shows up on your monitor at this point will not be automatically resized, so you are likely to see some visual corruption in the Interface window. To fix this, use the Spectator tablet app to toggle the camera off and on.
17. Verify that, throughout the above testing, the camera viewfinder overlay did not move or rotate. You can do this by looking around the viewfinder and ensuring it still looks "centered" in the "glass" part of the camera model.
18. In the Spectator tablet app, tap the words "HMD Preview" under "MONITOR SHOWS" to switch the slider to "HMD Preview".
19. Verify that the camera's viewfinder preview still shows up correctly.
20. Verify that your main monitor now shows the "READY - Put on your Headset" image
    - KNOWN ISSUE: This doesn't currently work. Instead, the monitor displays a frozen view of what the Spectator Camera last "saw". A similar issue occurs if you have the camera on and switch this switch to "HMD Preview" with the "Disable Preview" option checked.
21. Quit Interface using Tablet->File->Exit. Verify that you don't crash.
    - KNOWN ISSUE: Sometimes, you crash. See: https://trello.com/c/TcsggLra/189-fix-crash-that-occurs-when-closing-interface-after-creating-a-spectator-camera
